compiler/src/language_server/protocol.re

```diff
@@ -0,0 +1,203 @@
+[@deriving yojson]
+type command = {
+  title: string,
+  command: string,
+};
+
+// https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#traceValue
+// TODO: Implement custom yojson serde functions to convert to an enum
+[@deriving yojson]
+type trace_value = string; // 'off' | 'messages' | 'verbose';
+
+// https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocumentIdentifier
+[@deriving yojson]
+type text_document_identifier = {uri};
+
+// https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocumentSyncKind
+[@deriving (enum, yojson)]
+type text_document_sync_kind =
+  | [@value 0] No // This is `None` via the spec, but we don't want to collide with Option's `None`
+  | Full
+  | Incremental;
+
+let text_document_sync_kind_to_yojson = kind =>
+  text_document_sync_kind_to_enum(kind) |> [%to_yojson: int];
+let text_document_sync_kind_of_yojson = json =>
+  Result.bind(json |> [%of_yojson: int], value => {
+    switch (text_document_sync_kind_of_enum(value)) {
+    | Some(kind) => Ok(kind)
+    | None => Result.Error("Invalid enum value")
+    }
+  });
+
+// https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocumentItem
+[@deriving yojson({strict: false})]
+type text_document_item = {
+  uri,
+  languageId: string,
+  version: int,
+  text: string,
+};
+
+// https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#versionedTextDocumentIdentifier
+[@deriving yojson({strict: false})]
+type versioned_text_document_identifier = {
+  uri,
+  version: int,
+};
+
+// https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#requestMessage
+[@deriving yojson({strict: false})]
+type request_message = {
+  jsonrpc: version,
+  [@default None]
+  id: option(message_id),
+  method: string,
+  [@default None]
+  params: option(Yojson.Safe.t),
+};
+
+// https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#responseMessage
+[@deriving yojson({strict: false})]
+type response_message = {
+  jsonrpc: version,
+  id: option(message_id),
+  result: Yojson.Safe.t,
+};
+
+// https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#notificationMessage
+[@deriving yojson({strict: false})]
+type notification_message = {
+  jsonrpc: version,
+  method: string,
+  params: Yojson.Safe.t,
+};
+
+let version: version = "2.0";
+
+let header_prefix = "Content-Length: ";
+
+let sep = "\r\n\r\n";
+
+let request = (): result(request_message, string) => {
+  switch (input_line(stdin)) {
+  | exception exn => Error("Failed to read input")
+  | header when String_utils.starts_with(header, header_prefix) =>
+    let len =
+      String_utils.slice(
+        ~first=String.length(header_prefix),
+        ~last=-1 /* newline? */,
```

Comment:
@marcusroberts is this correct? We are skipping the newline? How does that work on Windows now that we are in binary mode?

Reply:
Windows in text mode used to be annoying because it stripped out the CR of the CRLF whereas Unix/Linux preserved it which is almost ironic...   Now Windows is in binary mode, we ignore the last CR (as input_line strips the LF only) on all platforms.  Comment updated.